Enable spellchecking in the EditText

Pluggi opened the pull request:
I thought #118 was a good idea, so here it is.
I only added a checkbox in the settings and tried to translate it in Serbian.

However, I was wondering if it would be a good idea to enable it automatically when there is no result that matches the query (e.g. when there is only the "Search on Google" thing), in order to be able to type faster when searching for something on Google/DDG.

saveman71 commented:
:+1: 

Your second point would be a good idea (as it would fulfill my request in #118), but I'm not sure how the keyboard would react if it was enabled on the fly while typing. I would say it deserves to be tried!

Pluggi commented:
I just tried it and it works perfectly well! At least with Google's keyboard.
There is no lags or anything like that.

My tests consisted in enabling the prediction when the length of the query was superior to 5.

Pluggi commented:
If someone could test with another keyboard (like Swiftkey for example), it would be great.

Here is a patch:

```diff
diff --git a/app/src/main/java/fr/neamar/kiss/MainActivity.java b/app/src/main/java/fr/neamar/kiss/MainActivity.java
index 5b43f0b..2e525e6 100644
--- a/app/src/main/java/fr/neamar/kiss/MainActivity.java
+++ b/app/src/main/java/fr/neamar/kiss/MainActivity.java
@@ -163,6 +163,16 @@ public class MainActivity extends ListActivity implements QueryInterface {
             public void onTextChanged(CharSequence s, int start, int before, int count) {
                 updateRecords(s.toString());
                 displayClearOnInput();
+
+                if (s.length() > 5) {
+                    searchEditText.setInputType(InputType.TYPE_CLASS_TEXT |
+                            InputType.TYPE_TEXT_FLAG_AUTO_CORRECT);
+                } else {
+                    searchEditText.setInputType(InputType.TYPE_CLASS_TEXT |
+                            InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD |
+                            InputType.TYPE_TEXT_FLAG_NO_SUGGESTIONS);
+                }
+
             }
         });
```

saveman71 commented:
Just for info if you want to test your patch, [Swiftkey](https://play.google.com/store/apps/details?id=com.touchtype.swiftkey&hl=en) is now free. (I can't test your patch, I am no android developer)

Pluggi commented:
I didn't know ! Thank you, I will try and test it tonight.

Pluggi commented:
Ok, so I've been thoroughly testing for the past hour and I think it's all good now !

However maybe we could find a better wording for @string/autospellcheck_dialog_message, but I am not able to come up with anything better for now.

Neamar commented:
Got a crash when setting a new value (on SettingActivity):


```
    Process: fr.neamar.kiss, PID: 29619
    java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Integer
            at android.app.SharedPreferencesImpl.getInt(SharedPreferencesImpl.java:239)
            at fr.neamar.kiss.SettingsActivity.fixSummaries(SettingsActivity.java:44)
            at fr.neamar.kiss.SettingsActivity.onSharedPreferenceChanged(SettingsActivity.java:26)
            at android.app.SharedPreferencesImpl$EditorImpl.notifyListeners(SharedPreferencesImpl.java:476)
            at android.app.SharedPreferencesImpl$EditorImpl.apply(SharedPreferencesImpl.java:384)
            at android.preference.Preference.tryCommit(Preference.java:1411)
            at android.preference.Preference.persistString(Preference.java:1444)
            at android.preference.EditTextPreference.setText(EditTextPreference.java:92)
            at android.preference.EditTextPreference.onDialogClosed(EditTextPreference.java:146)
            at android.preference.DialogPreference.onDismiss(DialogPreference.java:398)
            at android.app.Dialog$ListenersHandler.handleMessage(Dialog.java:1263)
            at android.os.Handler.dispatchMessage(Handler.java:102)
            at android.os.Looper.loop(Looper.java:135)
            at android.app.ActivityThread.main(ActivityThread.java:5291)
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
```

Neamar commented:
(actually, this killed my KISS, I can't start it anymore :p)

Neamar commented:
And finally, at least on Swiftkey, predictions are always displayed no matter what.

That's actually a known swiftkey bug, see #44, requiring to use PassswordVisible -- that's the only way to disable suggestions :(

Pluggi commented:
I thought I fixed the cast issue, but ok...

You're right about Swiftkey, the thing is that I removed the password visible input type because this was changing the font of the EditText and I didn't recheck with Swiftkey after removing it.

saveman71 commented:
Up!

Any updates about this @Pluggi

Pluggi commented:
Still on holidays, sorry. But feel free to contribute !

Neamar commented:
Sorry @Pluggi looks like I broke it again :(

Pluggi commented:
This should be ok now.

There may be some rewording to make in the preferences.

Neamar commented on app/src/main/java/fr/neamar/kiss/MainActivity.java:
```diff
@@ -76,6 +77,16 @@
+                                              InputType.TYPE_TEXT_FLAG_AUTO_CORRECT;
+
+    private final int spellcheckDisabledType = InputType.TYPE_CLASS_TEXT |
+                                               InputType.TYPE_TEXT_FLAG_NO_SUGGESTIONS;
```
Swiftkey displays suggestions even with this flag, we'll need to use the VISIBLE_PASSWORD flag.

Neamar replied:
```java
    private final int spellcheckDisabledType = InputType.TYPE_CLASS_TEXT |
                                               InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD;
```

Neamar commented on app/src/main/java/fr/neamar/kiss/MainActivity.java:
```diff
@@ -297,6 +310,16 @@ protected void onResume() {
+            searchEditText.setInputType(spellcheckEnabledType);
+        }
+        else {
+            searchEditText.setInputType(spellcheckDisabledType);
```
hum, thinking on this: we don't really need a `else`, since the default inputType (from the xml layout) is already correct.

The only case this could be useful is right after changing theme, and there's no need to set this on all other case.

Best thing i can think of is reusing the `require-layout-update` flag when switching the toggle.

That way, we can `setInputType()` on onCreate() once, and only if it's required \o/

Pluggi commented:
> hum, thinking on this: we don't really need a else, since the default inputType (from the xml layout) is already correct.
>
> The only case this could be useful is right after changing theme, and there's no need to set this on all other case.

This is also useful if the user toggles the spellcheck in the settings, because OnCreate is not called after going back from the settings.

Pluggi commented:
Is this ok now ?

saveman71 commented:
yay !
